lib/collection/src/shards/transfer/driver.rs

```diff
@@ -97,97 +101,53 @@ pub async fn transfer_shard(
+pub fn transfer_shard_fallback_default(
     mut transfer_config: ShardTransfer,
-    progress: Arc<Mutex<TransferTaskProgress>>,
-    shard_holder: Arc<LockedShardHolder>,
-    remote_shard: RemoteShard,
     consensus: &dyn ShardTransferConsensus,
-    collection_id: CollectionId,
     collection_name: &str,
-    channel_service: ChannelService,
-    snapshots_path: &Path,
-    temp_dir: &Path,
-) -> CollectionResult<()> {
-    let shard_id = transfer_config.shard_id;
-
-    // Fall back to stream records specifically
-    // TODO: fall back to default as specified in configuration
-    transfer_config
+    fallback_method: ShardTransferMethod,
+) -> CollectionResult<bool> {
+    // Do not attempt to fall back to the same method
+    if transfer_config
         .method
-        .replace(ShardTransferMethod::StreamRecords);
-
-    // Temporary hack for remote shard to get from PartialSnapshot back into Partial state
-    // To support stream records transfer, the remote shard must be in Partial state
+        .map_or(false, |method| method == fallback_method)
     {
-        let shard_holder_read = shard_holder.read().await;
-        let transferring_shard = shard_holder_read.get_shard(&shard_id);
-        let Some(replica_set) = transferring_shard else {
-            return Err(CollectionError::service_error(format!(
-                "Shard {shard_id} cannot be queue proxied because it does not exist"
-            )));
-        };
-        let shard_state = replica_set.peer_state(&transfer_config.to);
-
-        // Only switch if not in partial already
-        if shard_state != Some(ReplicaState::Partial) {
-            log::trace!("Shard {shard_id} diff recovered on {} for diff transfer, switching into next stage through consensus", remote_shard.peer_id);
-            consensus
-                .snapshot_recovered_switch_to_partial_confirm_remote(
-                    &transfer_config,
-                    collection_name,
-                    &remote_shard,
-                )
-                .await
-                .map_err(|err| {
-                    crate::operations::types::CollectionError::service_error(format!(
-                        "Can't switch shard {shard_id} to Partial state after diff transfer: {err}"
-                    ))
-                })?;
-        }
+        log::warn!("Failed shard transfer not falling back, because it would use the same transfer method: {fallback_method:?}");
+        return Ok(false);
     }
 
-    transfer_shard(
-        transfer_config,
-        progress,
-        shard_holder,
-        consensus,
-        collection_id,
-        collection_name,
-        channel_service,
-        snapshots_path,
-        temp_dir,
-    )
-    .await
+    // Propose to restart transfer with a different method
+    transfer_config.method.replace(fallback_method);
+    consensus.restart_shard_transfer(transfer_config, collection_name.into())?;
```

Comment:
We might want to confirm consensus actually picks this up. Thoughts?

Reply:
We have to check how restarting interacts with `sync_local_state`. If it will be eventually retried/cleaned up in `sync_local_state`, then this is, _probably_, fine. If not, then I think we either have to wait for consensus (with timeout error; as usual) or fix `sync_local_state`, IMO.

Reply:
I did think about this, but it proved very hard to do in this place. We don't have infrastructure in place here to wait on shard transfer states, nor can we access the `on_apply` callbacks in the consensus manager here.

I'll see what I can do to improve upon this.

Reply:
I found out we actually do have access to consensus. I've simply used our existing `propose_consensus_op_with_await` and added a loop for 3 retries. That should be good enough.

If retrying like this fails 3 times the failure is propagated to the current transfer, which means it will be aborted.